library/foreman_provisioner.py

```diff
@@ -0,0 +1,291 @@
+
+    def bmc(self, host_id, command):
+        """
+        execute a command through the BMC plugin (on/off/restart/shutdown/etc)
+        :param host_id: the id or name of the host as it is listed in foreman
+        :param command: the command to send through the BMC plugin, supported
+        commands: 'status', 'on', 'off', 'cycle', 'reset', 'soft'
+        """
+        request_url = '{0}/{1}/{2}/power'.format(self.url, self.default_uri, host_id)
+        command = json.dumps({'power_action': command})
+        response = self.session.put(request_url, data=command, verify=False)
+        #TODO(tkammer): add verification that the BMC command was issued
+
+    def ipmi(self, host_id, command):
+        """
+        execute a command through the ipmitool
+        :param host_id: the ipmi id of the host
+        :param command: the command to send through the ipmitool
+        commands: 'status', 'on', 'off', 'cycle', 'reset', 'soft' # TBD
+        """
+        command = "ipmitool -I lanplus -H {0} -U ADMIN -P ADMIN chassis power {1}".format(host_id, command)
+        response = os.system(command)
+
+    def _validate_bmc(self, host_id):
+        """
+        This method validate that there is at least one BMC on the given host
+        :param host_id: the id or name of the host as it is listed in foreman
+        """
+        request_url = '{0}/{1}/{2}/interfaces'.format(
+            self.url, self.default_uri, host_id)
+        response = self.session.get(request_url, verify=False)
+        body = response.json()
+        found_bmc = False
+        for interface in body['results']:
+            if "BMC" in interface['type']:
+                found_bmc = True
+                break
+        return found_bmc
+
+    def provision(self, host_id, mgmt_strategy, mgmt_action,
+                  wait_for_host=True):
+        """
+        This method rebuilds a machine, doing so by running get_host and bmc.
+        :param host_id: the name or ID of the host we wish to rebuild
+        :param mgmt_strategy: the way we wish to reboot the machine
+        (i.e: foreman, ipmi, etc)
+        :param mgmt_action: the action we wish to use with the strategy
+        (e.g: cycle, reset, etc)
+        :param wait_for_host: whether the function will wait until host has
+        finished rebuilding before exiting.
+        :raises: KeyError if BMC hasn't been found on the given host
+                 Exception in case of machine could not be reached after
+                 rebuild
+        """
+        if not self._validate_bmc(host_id):
+            raise KeyError("BMC not found on {}".format(host_id))
+        building_host = self.get_host(host_id)
+        self.set_build_on_host(host_id, True)
+        if mgmt_strategy == 'foreman':
+            self.bmc(host_id, mgmt_action)
+        elif mgmt_strategy == 'ipmi':
+            host = "{0}.{1}".format(building_host.get('interfaces')[0].get('name'),building_host.get('domain_name'))
```

Comment:
convention is to have a "space" between "," and the next value so: 
building_host.get('interfaces')[0].get('name'),building_host.get('domain_name')
should be 
building_host.get('interfaces')[0].get('name'), building_host.get('domain_name')